### PR TITLE
return specific error types for sub-IFDs

### PR DESCRIPTION
**Pull request by mpl**
Context: https://code.google.com/p/camlistore/issues/detail?id=493

We have some Camlistore users reporting that they have pictures on which we choke. It looks like goexif is rightly having a problem processing a badly formed tiff tag type in at least the first case, and that's fine.

However, I think they do have a point when they compare the result with some other tools that do report the error ("Warning: Bad InteropIFD directory" from exiftool), but still give whatever useful info was extracted from the (broken) file.

So I'd like to propose a mechanism to return some specific error types, so that we know when it's still possible to use whatever was decoded by goexif in some cases. Something like:

```
ex, err := exif.Decode(f)
if err != nil {
    if !exif.IsGPSInfoError(err) && !exif.IsInteroperabilityError(err) {
        log.Fatalf("exif.Decode: %v", err)
    }       
}
// else try and use ex anyway to get e.g. DateTime info.
```

I've just toyed with this idea and this commit is maybe a bit gross. If you're ok with the concept then I'll try and think of a better implementation.

### CODE REVIEW

**Comment by rwcarlsen** [comment]
This is actually something that has been on my (far too long) todo list for a while - a best effort decode when data is corrupt.  I'd be happy for such changes to be worked on by someone other than me ;-)

**Comment by mpl** [comment]
here's another approach, what do you think?

(I'll squash all the commits together eventually).

**Comment by rwcarlsen** [comment]
It looks fine.  The tiff package probably could use some modifications to do a best-effort decode.  Right now, as soon as it runs into any error, it returns nil up the stack regardless of what had been (and could have been) decoded correctly.  If an IFD offset is bad, then the decoding has to abort, but could still return prev IFD's.  Also, because tags within an IFD are fixed size, if a single tag is corrupt, decoding could continue for the rest of the IFD.

**Comment by mpl** [comment]
Ok, I've squashed the commits and added some doc.

Just to be sure though: you're saying that if an IFD offset is bad the decoding should abort. So does that mean that I should analyse an error returned by e.g. loadSubDir(x, ExifIFDPointer, exifFields) to make sure we don't have an offset error before letting loadSubDir(x, GPSInfoIFDPointer, gpsFields) run?

I understand your best-effort suggestions but I propose to try and address them in subsequent commits (as they're lower priority for me). Can't make any promise but I think I'll be able to find some time and give it a go if you haven't got around to it. I'll also open an issue about it the Camlistore issue tracker, as I think it could motivate some others to have a go at it.
Sounds good?

**Comment by mpl** [comment]
Just to be sure, is there anything you need from me about that change?

**Comment by rwcarlsen** [comment]
I'm busy with research now.  I added you as a committer - be good.

**Comment by mpl** [comment]
ack, thanks. I'll ask the other camli contributors to have a look at that
one before committing anyway.

**Comment by fawick** [comment]
> LGTM. Maybe the name of the erroneous field should be featured more prominently in the error string?

Done in loadSubDir

File: exif/exif.go [hunk 1]
```diff
@@ -62,6 +62,70 @@ func RegisterParsers(ps ...Parser) {
...
+		return isInterop
+	}
+	return false
+}
+
+const (
+	loadExif             = iota
+	loadGPS              = iota
+	loadInteroperability = iota
+)
```
Comment (saljam) [inline exif/exif.go:117]: You could just use implicit iotas here.

```
const (
    loadExif = iota
    loadGPS
    loadInteroperability
)
```

**Comment by saljam** [comment]
But that's just nitpicking. LGTM.

File: exif/exif.go [hunk 2]
```diff
@@ -62,6 +62,70 @@ func RegisterParsers(ps ...Parser) {
...
+	if tErr, ok := err.(tiffError); ok {
+		_, isInterop := tErr.stageMsg[loadInteroperability]
+		return isInterop
+	}
+	return false
+}
+
+const (
+	loadExif             = iota
+	loadGPS              = iota
```
Comment (bradfitz) [inline exif/exif.go]: yeah, drop the iotas.  And I'd make these have a type so it's obvious that they're a specific type of error code.

type somethingError int

const (
    loadExif somethingError = iota
    loadGPS
    loadFoo
)

File: exif/exif.go [hunk 3]
```diff
@@ -62,6 +62,70 @@ func RegisterParsers(ps ...Parser) {
...
+}
+
+// IsCriticalError, given the error returned by Decode, returns whether the
+// returned *Exif may contain usable information.
+func IsCriticalError(err error) bool {
+	_, ok := err.(tiffError)
+	return !ok
+}
+
+// IsExifError returns whether the error happened while decoding the EXIF
```
Comment (bradfitz) [inline exif/exif.go]: "reports whether", not "returns whether".  As Rob says: "It always returns."

But "reports whether" is Go style nowadays.

File: exif/exif.go [hunk 4]
```diff
@@ -62,6 +62,70 @@ func RegisterParsers(ps ...Parser) {
 
 type parser struct{}
 
+type tiffError struct {
+	// TODO(mpl): maybe just make tiffError type be the map directly?
+	stageMsg map[int]string
+}
+
+func (err tiffError) Error() string {
```
Comment (bradfitz) [inline exif/exif.go]: don't name the receiver "err". that makes it look like an error. just do "te".

File: exif/exif.go [hunk 5]
```diff
@@ -117,6 +191,8 @@ type Exif struct {
 // decoded, each registered parser is called (in order of registration). If
 // one parser returns an error, decoding terminates and the remaining
 // parsers are not called.
+// The error can be inspected with functions such as IsCriticalError to
+// determine whether the returned object might still be exploitable.
```
Comment (bradfitz) [inline exif/exif.go]: exploitable?

File: exif/exif.go (hunk 1 above)
Comment (mpl) [inline exif/exif.go:117]: Done.

File: exif/exif.go (hunk 2 above)
Comment (mpl) [inline exif/exif.go]: Done.

File: exif/exif.go (hunk 3 above)
Comment (mpl) [inline exif/exif.go]: Done.

File: exif/exif.go (hunk 4 above)
Comment (mpl) [inline exif/exif.go]: Done.

File: exif/exif.go (hunk 5 above)
Comment (mpl) [inline exif/exif.go]: s/exploitable/usable/

**Comment by mpl** [comment]
I'll squash and merge that PR tomorrow unless somebody objects.

**Pull request merged**